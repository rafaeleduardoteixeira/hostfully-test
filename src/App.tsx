import React from 'react';
import { createBrowserRouter, RouterProvider } from 'react-router-dom';
import createdRoutes from './constants/router';
import { Header } from './components/Header/Header';
import { AppContainer } from './App.styles';

function App() {
  return (
    <AppContainer>
      <Header />
      <RouterProvider router={createdRoutes} />
    </AppContainer>
  );
}

export default App;
