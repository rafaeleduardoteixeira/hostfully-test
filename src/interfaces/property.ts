export interface Property {
  id: string;
  name: string;
  description: string;
  city: string;
  country: string;
  purpose: string;
  price: number;
  images: string[];
  position: {
    latitude: number;
    longitude: number;
  };
  rentals?: {
    user: string;
    startDate: string;
    endDate: string;
  }[];
}
