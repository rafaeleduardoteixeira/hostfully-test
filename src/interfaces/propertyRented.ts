export interface PropertyRented {
  id: string;
  propertyId: string;
  startDate: string;
  endDate: string;
}
