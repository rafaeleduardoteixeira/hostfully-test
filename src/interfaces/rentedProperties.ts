export interface RentedProperties {
  id: string;
  propertyId: string;
  propertyImages?: string[];
  propertyName?: string;
  name: string;
  email: string;
  checkIn: Date | null;
  checkOut: Date | null;
}
