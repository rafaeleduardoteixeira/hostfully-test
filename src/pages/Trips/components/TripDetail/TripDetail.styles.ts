import styled from 'styled-components';
import { media } from '../../../../themes';

export const Container = styled.div`
  height: auto;
  display: flex;
  flex-direction: column;
  gap: 20px;
  padding-bottom: 20px;
  width: 100%;
`;

export const TripDetailTitle = styled.h1`
  font-size: 1.5rem;
  font-weight: 500;

  text-align: center;
`;

export const DataContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;
`;

export const ButtonContainer = styled.div`
  display: flex;
  justify-content: center;
  gap: 20px;

  @media ${media.mobile} {
    flex-direction: column;
  }
`;

interface TripDetalButtonProps {
  $backgroundColor?: string;
  $backgroundColorHover?: string;
  color?: string;
}

export const TripDetalButton = styled.button<TripDetalButtonProps>`
  width: 100%;
  padding: 10px 20px;
  border: none;
  border-radius: 10px;
  background-color: ${({ $backgroundColor }) => $backgroundColor || 'hsl(200, 10%, 90%)'};
  color: ${({ color }) => color || 'hsl(200, 10%, 20%)'};
  font-size: 1rem;
  font-weight: 500;
  cursor: pointer;
  transition: background-color 0.3s;

  &:hover {
    background-color: ${({ $backgroundColorHover }) => $backgroundColorHover || 'hsl(200, 10%, 95%)'};
  }
`;
