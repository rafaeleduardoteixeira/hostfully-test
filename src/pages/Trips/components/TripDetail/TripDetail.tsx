import { ChangeEvent, useContext, useState } from 'react';
import { Input } from '../../../../components/Input/Input';
import { RentedProperties } from '../../../../interfaces/rentedProperties';
import { colors } from '../../../../themes';
import { Container, TripDetailTitle, DataContainer, ButtonContainer, TripDetalButton } from './TripDetail.styles';
import { InputDateRange } from '../../../../components/InputDateRange/InputDateRange';
import moment from 'moment';
import { AppContext, AppContextType } from '../../../../Context/App.context';
import { set } from 'react-datepicker/dist/date_utils';

interface TripDetailProps {
  rentedProperty: RentedProperties;
  handleEditTrip: (rentedProperty: RentedProperties) => void;
  handleDeleteTrip: (rentedProperty: RentedProperties) => void;
  handleCancel: () => void;
}

export const TripDetail = ({ handleEditTrip, handleDeleteTrip, handleCancel, rentedProperty }: TripDetailProps) => {
  const [checkIn, setCheckIn] = useState(rentedProperty.checkIn);
  const [checkOut, setCheckOut] = useState(rentedProperty.checkOut);
  const [error, setError] = useState<string>('');

  const { getPropertyRentedDates } = useContext(AppContext) as AppContextType;
  const rentedDates = getPropertyRentedDates(rentedProperty.propertyId);

  const handleCheckRentedDates = (checkIn: Date, checkOut: Date) => {
    // Create to prevent that user select a date that is already rented. For instance if the user select 23/07/2024 to 27/07/2024 and the property is already rented from 24/07/2024 to 26/07/2024, the user can't select this date. I set the checkIn date the value of checkOut date.
    if (checkIn && checkOut) {
      return rentedDates?.some((date) => {
        if (date.start && date.end) {
          if (
            (checkIn >= date.start && checkIn <= date.end) ||
            (checkOut >= date.start && checkOut <= date.end) ||
            (checkIn <= date.start && checkOut >= date.end)
          )
            return true;
        }
        return false;
      });
    }
    return false;
  };

  const handleChange = (value: Date[] | null) => {
    if (value) {
      if (handleCheckRentedDates(value[0], value[1])) {
        setCheckIn(null);
        setCheckOut(null);
        return;
      }
      setCheckIn(value[0]);
      setCheckOut(value[1]);
      return;
    }
    setCheckIn(null);
    setCheckOut(null);
  };

  const validateDates = () => {
    if (!checkIn || !checkOut) {
      setError('Invalid date');
      return true;
    }
    setError('');
    return false;
  };

  const handleEdit = (rentedProperty: RentedProperties) => {
    if (validateDates()) return;

    handleEditTrip({
      ...rentedProperty,
      checkIn,
      checkOut,
    });
  };

  return (
    <Container>
      <TripDetailTitle>Trip Detail</TripDetailTitle>
      <Input label="Name" value={rentedProperty.name} disabled={true} />
      <Input label="Property" value={rentedProperty.propertyName} disabled={true} />
      <DataContainer>
        <InputDateRange
          bookedDates={rentedDates}
          dateStart={checkIn ? moment(checkIn, 'America/Los_Angeles').toDate() : undefined}
          dateEnd={checkOut ? moment(checkOut, 'America/Los_Angeles').toDate() : undefined}
          error={error}
          onChange={handleChange}
        />
      </DataContainer>
      <ButtonContainer>
        <TripDetalButton
          onClick={() => handleEdit(rentedProperty)}
          $backgroundColor={colors.hostfullyPrimaryColor}
          $backgroundColorHover={colors.hostfullyPrimaryColorHover}
          color="white"
        >
          Edit
        </TripDetalButton>
      </ButtonContainer>
      <ButtonContainer>
        <TripDetalButton
          onClick={() => handleDeleteTrip(rentedProperty)}
          $backgroundColor={colors.red}
          $backgroundColorHover={colors.redDarker}
          color="white"
        >
          Delete
        </TripDetalButton>
        <TripDetalButton onClick={handleCancel}>Cancel</TripDetalButton>
      </ButtonContainer>
    </Container>
  );
};
