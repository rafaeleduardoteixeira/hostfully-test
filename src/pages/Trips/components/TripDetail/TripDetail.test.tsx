import React, { act } from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { TripDetail } from './TripDetail';
import { AppContext } from '../../../../Context/App.context';
import { RentedProperties } from '../../../../interfaces/rentedProperties';
import moment from 'moment';

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}));

const rendedProperties: RentedProperties = {
  id: '1',
  propertyId: '1',
  propertyImages: ['image'],
  propertyName: 'Property 1',
  name: 'Name 1',
  email: 'Email 1',
  checkIn: new Date(),
  checkOut: new Date(),
};

describe('TripDetail', () => {
  test('Snapshot test', () => {
    const { asFragment } = render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <TripDetail
          rentedProperty={rendedProperties}
          handleEditTrip={jest.fn()}
          handleDeleteTrip={jest.fn()}
          handleCancel={jest.fn()}
        />
      </AppContext.Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('Should render TripDetail', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <TripDetail
          rentedProperty={rendedProperties}
          handleEditTrip={jest.fn()}
          handleDeleteTrip={jest.fn()}
          handleCancel={jest.fn()}
        />
      </AppContext.Provider>,
    );
    const TripDetailElement = screen.getByTestId('TripDetail');
    expect(TripDetailElement).toBeInTheDocument();
  });
  test('Should edit a trip', () => {
    const handleEditTrip = jest.fn();
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <TripDetail
          rentedProperty={rendedProperties}
          handleEditTrip={handleEditTrip}
          handleDeleteTrip={jest.fn()}
          handleCancel={jest.fn()}
        />
      </AppContext.Provider>,
    );
    const editButton = screen.getByText('Edit');
    editButton.click();
    expect(handleEditTrip).toHaveBeenCalled();
  });
  test('Should delete a trip', () => {
    const handleDeleteTrip = jest.fn();
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <TripDetail
          rentedProperty={rendedProperties}
          handleEditTrip={jest.fn()}
          handleDeleteTrip={handleDeleteTrip}
          handleCancel={jest.fn()}
        />
      </AppContext.Provider>,
    );
    const deleteButton = screen.getByText('Delete');
    deleteButton.click();
    expect(handleDeleteTrip).toHaveBeenCalled();
  });
  test('Should cancel a trip', () => {
    const handleCancel = jest.fn();
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <TripDetail
          rentedProperty={rendedProperties}
          handleEditTrip={jest.fn()}
          handleDeleteTrip={jest.fn()}
          handleCancel={handleCancel}
        />
      </AppContext.Provider>,
    );
    const cancelButton = screen.getByText('Cancel');
    cancelButton.click();
    expect(handleCancel).toHaveBeenCalled();
  });
  test('Should try to edit with a null date', () => {
    const handleEditTrip = jest.fn();
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: jest.fn(),
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <TripDetail
          rentedProperty={{ ...rendedProperties, checkIn: null, checkOut: null }}
          handleEditTrip={handleEditTrip}
          handleDeleteTrip={jest.fn()}
          handleCancel={jest.fn()}
        />
      </AppContext.Provider>,
    );
    const editButton = screen.getByText('Edit');
    editButton.click();
    expect(handleEditTrip).not.toHaveBeenCalled();
  });
});
