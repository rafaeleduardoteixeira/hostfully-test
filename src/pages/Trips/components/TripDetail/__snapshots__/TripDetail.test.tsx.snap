// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TripDetail Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="sc-fQpRED ibyllv"
    data-testid="TripDetail"
  >
    <h1
      class="sc-dsLQwm emstNv"
    >
      Trip Detail
    </h1>
    <div
      class="sc-braxZu jqDACe"
    >
      <label
        class="sc-gJhJTp MduwH"
      >
        Name
      </label>
      <input
        class="sc-elDIKY hDaMyU"
        data-testid="Input"
        disabled=""
        label="Name"
        value="Name 1"
      />
    </div>
    <div
      class="sc-braxZu jqDACe"
    >
      <label
        class="sc-gJhJTp MduwH"
      >
        Property
      </label>
      <input
        class="sc-elDIKY hDaMyU"
        data-testid="Input"
        disabled=""
        label="Property"
        value="Property 1"
      />
    </div>
    <div
      class="sc-iKTcqh cebKEX"
    >
      <div
        class="sc-iCZwEW gtdYpf"
        error=""
      >
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container react-datepicker__view-calendar-icon"
          >
            <img
              alt="calendar"
              class=" react-datepicker__calendar-icon "
              src="/icons/calendar.svg"
            />
            <input
              class="date-picker"
              placeholder="Select a date to book"
              type="text"
              value="July 24, 2024 - July 24, 2024"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-gnpbhQ esMGKT"
    >
      <button
        class="sc-la-DxNn iSrhNw"
        color="white"
      >
        Edit
      </button>
    </div>
    <div
      class="sc-gnpbhQ esMGKT"
    >
      <button
        class="sc-la-DxNn hTaAUK"
        color="white"
      >
        Delete
      </button>
      <button
        class="sc-la-DxNn gnKfZG"
      >
        Cancel
      </button>
    </div>
  </div>
</DocumentFragment>
`;
